In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("Traffic_Police_Data.csv")

See the data

In [3]:
df.head()

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   
2  1/23/2005     23:15           NaN             M          1972.0   
3  2/20/2005     17:15           NaN             M          1986.0   
4  3/14/2005     10:00           NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

So now we can see the top 5 rows with headings from the data.

In [5]:
df.shape

(65535, 15)

Now we can see that there are total (65535) rows and (15) columns.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           65535 non-null  object 
 1   stop_time           65535 non-null  object 
 2   country_name        0 non-null      float64
 3   driver_gender       61474 non-null  object 
 4   driver_age_raw      61481 non-null  float64
 5   driver_age          61228 non-null  float64
 6   driver_race         61475 non-null  object 
 7   violation_raw       61475 non-null  object 
 8   violation           61475 non-null  object 
 9   search_conducted    65535 non-null  bool   
 10  search_type         2479 non-null   object 
 11  stop_outcome        61475 non-null  object 
 12  is_arrested         61475 non-null  object 
 13  stop_duration       61475 non-null  object 
 14  drugs_related_stop  65535 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 6.6+

Here we can see all information about the data like how many values are there and what kind of data type.

In [8]:
df.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

Now we can see how many null value are there in each column. If there are a column with maximum null values we don't use that in our further analysis so we just drop that column from there

In [9]:
df.drop(columns = 'country_name',inplace=True)

country_name has all blank rows so I just drop that column.

In [11]:
df.shape

(65535, 14)

Now again we can see the data and now there are only 14 columns.

In [12]:
df[df.violation == 'Speeding'].driver_gender.value_counts()

M    25517
F    11686
Name: driver_gender, dtype: int64

I can see that there is a column for male and female, so I divide the data in those two categories.
M = 25517
F = 11686

In [13]:
df.search_conducted.value_counts()

False    63056
True      2479
Name: search_conducted, dtype: int64

Here we can see that howmany times there are search conducted.

In [14]:
df.groupby('driver_gender').search_conducted.sum()

driver_gender
F     366
M    2113
Name: search_conducted, dtype: int64

with the group by method I divide the search conducted into male and female.

In [15]:
df['stop_duration'].value_counts()

0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: stop_duration, dtype: int64

Here we can see that how many times police stop a person till what time range?

In [21]:
v = df.groupby('violation')
v

Here I just create a new variable and groupby the column violation

In [24]:
for violation , violation_df in v:
    print(violation)


Equipment
Moving violation
Other
Registration/plates
Seat belt
Speeding


Here I use the for loop for printing all the types of violations.

In [30]:
v.max().driver_age

C:\Users\Aman\AppData\Local\Temp\ipykernel_17908\4053235499.py:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.max is deprecated. In a future version, a TypeError will be raised. Before calling .max, select only columns which should be valid for the function.
  v.max().driver_age


violation
Equipment              81.0
Moving violation       86.0
Other                  86.0
Registration/plates    74.0
Seat belt              42.0
Speeding               88.0
Name: driver_age, dtype: float64

Here we can see the maximum age for each type of violation.

In [29]:
v.min().driver_age

C:\Users\Aman\AppData\Local\Temp\ipykernel_17908\1820596496.py:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.min is deprecated. In a future version, a TypeError will be raised. Before calling .min, select only columns which should be valid for the function.
  v.min().driver_age


violation
Equipment              16.0
Moving violation       15.0
Other                  16.0
Registration/plates    16.0
Seat belt              23.0
Speeding               15.0
Name: driver_age, dtype: float64

Here we can see the minimum age for each type of violation.

In [32]:
v.mean().driver_age

violation
Equipment              31.682957
Moving violation       36.736443
Other                  40.362381
Registration/plates    32.656696
Seat belt              30.333333
Speeding               33.262581
Name: driver_age, dtype: float64

Now this is the mean value of age for each type of violation.

In [33]:
v.describe().driver_age

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0

Here we can see all the various statistics like mean, standarn deviation and all othere distributions.